In [130]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Import tập dữ liệu

Sau đó chúng ta import tập dữ liệu.

In [137]:
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-2].values
y = dataset.iloc[:, -1].values

In [15]:
import re
import os
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
with open(os.path.join("aclImdb","train", "neg",'10040_2.txt'), "r", encoding="utf-8") as f:
    print(f.read())

def clean_text(text):
    """
    Applies some pre-processing on the given text.

    Steps :
    - Removing HTML tags
    - Removing punctuation
    - Lowering text
    """
    
    # loại bỏ HTML tag
    text = re.sub(r'<.*?>', '', text)
    
    # CODE CỦA BẠN Ở ĐÂY
    # loại bỏ các ký tự [\], ['] và ["] bằng phương thức resub: 
    # CODE CỦA BẠN Ở ĐÂY
    text = re.sub(r"[\]", '', text)

    text = re.sub(r"[']", '', text)  

    text = re.sub(r'["]', '', text) 
    
    # chuyển đổi văn bản thành chữ thường
    text = text.strip().lower()
    
    # thay dấu chấm bằng dấu cách
    filters='!"\'#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n'
    translate_dict = dict((c, " ") for c in filters)
    translate_map = str.maketrans(translate_dict)
    text = text.translate(translate_map)

    return text

# Ví dụ
clean_text("<html>This is not a sentence.<\html>").split()

...for this movie defines a new low in Bollywood and has set the standard against which all c**p must now be compared.<br /><br />First off, the beginning did have elements of style....and if handled well, could have become a cult classic, a-la pulp fiction or a Desi desperado...but the plot (was there one?) begins to meander and at one point completely loses it.<br /><br />Throw in a deranged don with an obsession for English, a call center smart Alec, a femme fa tale who can don a bikini and a Saree with the same aplomb, a levitating, gravity defying hit-man and a cop with a hundred (or was it a thousand) black cat commandos on their trail....good ingredients in competent hands. But this is where I would like to ask the director: Sir, what were you smoking?<br /><br />Im sure this movie would be remembered in the annals of Bollywood film making - for what must never be done - insult the intelligence of the most brain dead of movie goers. <br /><br />Possibly the only redeeming featur

error: unterminated character set at position 0

## Mã hóa dữ liệu phân loại

In [139]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

[[0.0 0.0 1.0 165349.2 136897.8 471784.1]
 [1.0 0.0 0.0 162597.7 151377.59 443898.53]
 [0.0 1.0 0.0 153441.51 101145.55 407934.54]
 [0.0 0.0 1.0 144372.41 118671.85 383199.62]
 [0.0 1.0 0.0 142107.34 91391.77 366168.42]
 [0.0 0.0 1.0 131876.9 99814.71 362861.36]
 [1.0 0.0 0.0 134615.46 147198.87 127716.82]
 [0.0 1.0 0.0 130298.13 145530.06 323876.68]
 [0.0 0.0 1.0 120542.52 148718.95 311613.29]
 [1.0 0.0 0.0 123334.88 108679.17 304981.62]
 [0.0 1.0 0.0 101913.08 110594.11 229160.95]
 [1.0 0.0 0.0 100671.96 91790.61 249744.55]
 [0.0 1.0 0.0 93863.75 127320.38 249839.44]
 [1.0 0.0 0.0 91992.39 135495.07 252664.93]
 [0.0 1.0 0.0 119943.24 156547.42 256512.92]
 [0.0 0.0 1.0 114523.61 122616.84 261776.23]
 [1.0 0.0 0.0 78013.11 121597.55 264346.06]
 [0.0 0.0 1.0 94657.16 145077.58 282574.31]
 [0.0 1.0 0.0 91749.16 114175.79 294919.57]
 [0.0 0.0 1.0 86419.7 153514.11 0.0]
 [1.0 0.0 0.0 76253.86 113867.3 298664.47]
 [0.0 0.0 1.0 78389.47 153773.43 299737.29]
 [0.0 1.0 0.0 73994.56 122782.75 3

## Chia tập dữ liệu thành Training set và Test set

Sử dụng **train_test_split()** từ scikit-learning để thực hiện chia tách với **test_size = 0.2** của tập dữ liệu gốc và random_state = 0 để đảm bảo bạn sẽ có cùng tập dữ liệu khi thực hiện việc chia tách.

In [140]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

## Huấn luyện mô hình Hồi quy tuyến tính đa biến trên Training set

In [141]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression()

## Dự đoán kết quả Test set

In [142]:
y_pred = regressor.predict(X_test)
np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[103901.9  103282.38]
 [132763.06 144259.4 ]
 [133567.9  146121.95]
 [ 72911.79  77798.83]
 [179627.93 191050.39]
 [115166.65 105008.31]
 [ 67113.58  81229.06]
 [ 98154.81  97483.56]
 [114756.12 110352.25]
 [169064.01 166187.94]]


## Đưa ra một dự đoán duy nhất (chẳng hạn: lợi nhuận của một startup với R&D Spend = 160000, Administration Spend = 130000, Marketing Spend = 300000 and State = 'California')

In [144]:
print(regressor.predict([[ 160000, 130000, 300000]]))

[181837.6]


Do đó, mô hình của chúng ta dự đoán rằng lợi nhuận của một startup ở California (đã chi 160000 USD cho R&D, 130000 USD cho Quản trị và 300000 cho Tiếp thị) là 181566,92 USD.

**Lưu ý quan trọng 1:** Lưu ý rằng tất cả các giá trị của các feature đều được nhập trong một cặp dấu ngoặc vuông kép. Đó là bởi vì phương thức "predict" luôn cần một mảng 2D là định dạng của các input. Và việc đặt các giá trị vào một cặp dấu ngoặc vuông kép làm cho input chính xác là một mảng 2D. Chỉ cần đặt:

$1, 0, 0, 160000, 130000, 300000 \rightarrow \textrm{scalars}$

$[1, 0, 0, 160000, 130000, 300000] \rightarrow \textrm{1D array}$

$[[1, 0, 0, 160000, 130000, 300000]] \rightarrow \textrm{2D array}$

**Lưu ý quan trọng 2:** Cũng lưu ý rằng trạng thái "California" không được nhập dưới dạng string trong cột cuối cùng mà là "1, 0, 0" trong ba cột đầu tiên. Đó là bởi vì phương thức dự đoán yêu cầu các giá trị one-hot coded của trạng thái và như chúng ta thấy ở hàng thứ hai của ma trận các feature X, "California" được mã hóa thành "1, 0, 0". Và hãy chú ý bao hàm các giá trị này trong ba cột đầu tiên chứ không phải ba cột cuối cùng, vì các biến giả luôn được tạo trong các cột đầu tiên.


## Nhận phương trình hồi quy tuyến tính cuối cùng với giá trị của các hệ số

In [33]:
print(regressor.coef_)
print(regressor.intercept_)

[ 8.66e+01 -8.73e+02  7.86e+02  7.73e-01  3.29e-02  3.66e-02]
42467.52924855596


Như vậy, phương trình mô hình hồi quy tuyến tính đa biến của chúng ta là:

$$\textrm{Profit} = 86.6 \times \textrm{Dummy State 1} - 873 \times \textrm{Dummy State 2} + 786 \times \textrm{Dummy State 3} + 0.773 \times \textrm{R&D Spend} + 0.0329 \times \textrm{Administration} + 0.0366 \times \textrm{Marketing Spend} + 42467.53$$

**Lưu ý quan trọng:** Để có được các hệ số này, chúng ta gọi thuộc tính "coef_" và "intercept_" từ đối tượng hồi quy. Các thuộc tính trong Python khác với các phương thức và thường trả về một giá trị đơn giản hoặc một mảng các giá trị.